### ISSUE TITLE
Migration from 1.0.2u to 3.1.1- Problem to transform RSA structure to EVP_PKEY

### ISSUE THREAD

**Issue by hololqq** [openssl/openssl#21238]
Hi everyone, I'm new to OpenSSL and we need to update openssl to the  latest 3.1.1. 
But the EVP_PKEY_assign_RSA() were deprecated so we need a way to transform the current RSA object into EVP_PKEY.
So I want to know which function I should use to finish this.

Here's the code I'm working.

```
RSA* pRsa = NULL;
EVP_PKEY* pKey = NULL;

// ... Previous code we've open a key file and read the data into the pRsa.
   pKey = EVP_PKEY_new();
   EVP_PKEY_assign_RSA( pKey, pRsa );
   pRsa = NULL;

   ret = X509_REQ_set_pubkey( pX509Req, pKey );
// .. Processing the pX509Req.
```

At the man page it suggested to use [OSSL_DECODER_CTX_new_for_pkey(3)] or [EVP_PKEY_fromdata(3)]. In my view the second method EVP_PKEY_fromdata should work but it still doesn't help to use the current data in the RSA object.

Thanks!

**Comment by paulidale** [comment]
Have you read the [migration guide](https://www.openssl.org/docs/man3.0/man7/migration_guide.html)?
You're probably seeking [EVP_PKEY_fromdata](https://www.openssl.org/docs/man3.0/man3/EVP_PKEY_fromdata.html).

**Comment by hololqq** [comment]
Yes I'm still looking at that guide and make tests with the fromData

**Comment by hololqq** [comment]
Hi @paulidale 
To convert a RSA* key to EVP_PKEY it's need to use fromData. But that one requires param from the RSA key( it's impossible to give me any key except RSA* structure). I tried to use RSA_get0_xxx to get the param. But these're also deprecated.
~~~
      if( bld == NULL
            || !OSSL_PARAM_BLD_push_BN(bld, "n", RSA_get0_n(pRsa) )
            || !OSSL_PARAM_BLD_push_uint(bld, "e", RSA_get0_e(pRsa) )
            || !OSSL_PARAM_BLD_push_BN(bld, "d", RSA_get0_d(pRsa) )
            || !OSSL_PARAM_BLD_push_BN(bld, "rsa-factor1", RSA_get0_p(pRsa) )
            || !OSSL_PARAM_BLD_push_BN(bld, "rsa-factor2", RSA_get0_q(pRsa) )
            || !OSSL_PARAM_BLD_push_BN(bld, "rsa-exponent1", RSA_get0_dmp1(pRsa) )
            || !OSSL_PARAM_BLD_push_BN(bld, "rsa-exponent2", RSA_get0_dmq1(pRsa) )
            || !OSSL_PARAM_BLD_push_BN(bld, "rsa-coefficient1", RSA_get0_iqmp(pRsa) )
            || (params = OSSL_PARAM_BLD_to_param(bld)) == NULL)
      {
         OSSL_PARAM_BLD_free(bld);
         return false;
      }
~~~

One thing is that the migration guide said the RSA_get0_xx can be replaced by EVP_PKEY_get_bn_param(3) etc, but they acquire EVP_PKEY* parameter not RSA*. That's crazy. 
But I found the deprecated functions could still be called when I build the dOpenSSL
~~~
In file included from ../../src/dopenssl/rsa.h:15,
                 from ../../src/dopenssl/rsa.c:12:
/openssl/rsa.h:239:37: note: declared here
 OSSL_DEPRECATEDIN_3_0 const BIGNUM *RSA_get0_e(const RSA *d);
                                     ^~~~~~~~~~
../../src/dopenssl/rsa.c:311:3: warning: 'RSA_get0_n' is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
   if ((n = BN_bn2hex(RSA_get0_n(rsa))) == NULL)
   ^~
~~~
As it returns error when I use it in other project
~~~
invalid use of incomplete type 'RSA' {aka 'struct rsa_st'}
             || !OSSL_PARAM_BLD_push_BN(bld, "n", pRsa->n)//RSA_get0_n(pRsa) )
~~~
So my question is, how can I call the deprecated function successfully as dOpenSSL did? I guess there's a macrio. Thanks.

**Comment by mattcaswell** [comment]
> To convert a RSA* key to EVP_PKEY it's need to use fromData. But that one requires param from the RSA key

*All* functions that operate on the `RSA` object are deprecated. You should seek to stop using the `RSA` object completely to avoid deprecation warnings (use EVP_PKEY instead). In other words don't construct an RSA first and then assign it to an EVP_PKEY. Instead you must construct an EVP_PKEY with the appropriate RSA values in it directly.

> it's impossible to give me any key except RSA* structure

This sounds like the source of your problem. Why is this impossible? If you only have an `RSA` object then you will inevitably have to call deprecated functions to use it.

> So my question is, how can I call the deprecated function successfully as dOpenSSL did? I guess there's a macro. Thanks.

I'm not familiar with dOpenSSL. But from a quick look it appears that it was written for use with 1.0.x OpenSSL. It would need significant work to upgrade it for use with OpenSSL 3. Most significantly the RSA structure was made opaque in 1.1.0 so that it is no longer possible to access the internal members of `RSA` directly as dOpenSSL appears to be doing.

**Comment by hololqq** [comment]
> 
Thanks for your reply.

The reason that I still use the RSA* structure is because we're using dOpenSSL to generate deterministic random key with char* we gave. We first generate a RSA key from dOpenSSL then process it. So it causes the problem that I can't convert the RSA* or reading param from it.

Yes the current dOpenSSL is written in 1.0.x. I made some API update to let it running with OpenSSL 3.1.1. So it's the phenomenon I found that that project could still use the deprecated RSA function. From OpenSSL code I saw the RSA* is still using internally so I'm guessing if it's still allow to call the deprecated API. 
Completely delete the RSA* method might affect the deterministic random generation functionalities of dOpenSSL so I'm looking for some warm way unless deprecating is need.

**Comment by mattcaswell** [comment]
> Yes the current dOpenSSL is written in 1.0.x. I made some API update to let it running with OpenSSL 3.1.1.

It sounds like you are doing the minimal changes required in dOpenSSL to get it to compile. Without changing dOpenSSL to not use the RSA structure you have no choice but to call deprecated functions. So you must live with the deprecation warnings for calls like `RSA_get0_n()`. Convert the RSA structure into an EVP_PKEY like you have done and restrict the deprecation warnings to that one area of code. Use EVP_PKEY everywhere else.

To make the deprecation warnings go away completely you need to do more work on dOpenSSL.

**Comment by hololqq** [comment]
OK..looks like that I need to reconfigure the entire dOpenSSL since ENGINE* are deprecated entire.

**Comment by nhorman** [comment]
marking as inactive, will be closed at the completion of 3.4 dev without further input

**Comment by t8m** [comment]
The question was answered. Closing.
